### PR TITLE
avoid deadlock when calling s3async callback, issue #1854

### PR DESCRIPTION
**Pull request by h0hmj**
Signed-off-by: h0hmj <h0hmjcn@gmail.com>

### What problem does this PR solve?

Issue Number: #1854 

Problem Summary:

### What is changed and how it works?

What's Changed:

add a work queue and a thread to handle async s3 request

How it Works:

move throttle check outside main loop, async s3 request will not block main loop

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by h0hmj** [comment]
recheck

File: src/common/s3_adapter.cpp [hunk 1]
```diff
@@ -332,6 +335,11 @@ int S3Adapter::PutObject(const Aws::String &key, const std::string &data) {
 */
 
 void S3Adapter::PutObjectAsync(std::shared_ptr<PutObjectAsyncContext> context) {
+    asyncRequestTP_.Enqueue(&S3Adapter::PutObjectAsyncHelper, this, context);
```
Comment (wuhongsong) [inline src/common/s3_adapter.cpp:338]: can cause excessive memory usage？

Comment (wu-hanqing) [reply src/common/s3_adapter.cpp:338]: I have the same question

Comment (h0hmj) [reply src/common/s3_adapter.cpp:338]: yes, i will change it, seems that just put callbacks into queue is enough.

**Comment by h0hmj** [comment]
let callbacks run in a separate thread rather than the event main loop to avoid deadlock.
num of callback ~= num of request (limited by inflight bytes)

File: src/common/s3_adapter.h [hunk 2]
```diff
@@ -332,6 +333,19 @@ class S3Adapter {
         std::condition_variable cond_;
     };
 
+    class AsyncCallbackThreadPool : public TaskThreadPool<> {
```
Comment (wu-hanqing) [inline src/common/s3_adapter.h:336]: it seems this derived class is unnecessary, we can add an extra parameter to control whether run all tasks when task pool stopped.

```
    virtual void ThreadFunc() {
        while (running_.load(std::memory_order_acquire)
|| (drain && !queue.empty())
) {
            Task task(Take());
            /* ThreadPool 退出的时候，queue 为空，那么会返回无效的 task */
            if (task) {
                task();
            }
        }
    }
```

File: src/common/s3_adapter.cpp [hunk 3]
```diff
@@ -340,13 +343,18 @@ void S3Adapter::PutObjectAsync(std::shared_ptr<PutObjectAsyncContext> context) {
...
-         originCallback](const std::shared_ptr<PutObjectAsyncContext>& ctx) {
+         originCallback](const std::shared_ptr<PutObjectAsyncContext> ctx) {
             inflightBytesThrottle_->OnComplete(ctx->bufferSize);
             ctx->cb = originCallback;
             ctx->cb(ctx);
         };
+    auto asyncCallback =
+        [this,
+         throttledCallback](const std::shared_ptr<PutObjectAsyncContext>& ctx) {
+            asyncCallbackTP_.Enqueue(std::move(throttledCallback), ctx);
```
Comment (wu-hanqing) [inline src/common/s3_adapter.cpp:356]: can we just do the Enqueue in L378 rather than wrap another call back?

Comment (h0hmj) [reply src/common/s3_adapter.cpp]: yes, i'll put it in handle function

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**